### DISCUSSION TITLE
Using Navigation for regions in the Shell when application starts

### DISCUSSION THREAD

**Discussion by angelotrivelli** [PrismLibrary/Prism#2586] [WPF] [answered]
I've been using View injection and it's worked out OK. But after seeing the Langunas pluralsight course, it seems like the way to go is to use Region Navigation. 

That works fine too and it's especially nice to associate view and viewmodels with `RegisterForNavigation<View,ViewModel>()` and then later on use `regionManager.RequestNavigate(viewNameString, viewModelString)` to get a view to show up in a region.

I started converting over my application to use Region Navigation, but ran into a problem with my MainWindow View (the "Shell"). What I got is a MainWindow with a few Regions in it (eg header, footer, sidebar...). At the start of the application, I want these to be populated with particular Views.

What I did was to inject `IRegionManager` into MainWindow, and then just tried to naively call `regionManager.RequestNavigate(myRegion,myView)` from my MainWindow contstructor. That didn't work, the view didn't show up. Turns out, it was because none of the regions existed yet.  I reproduced the problem starting from one of the examples ([17-BasicRegionNavigation](https://github.com/PrismLibrary/Prism-Samples-Wpf/tree/master/17-BasicRegionNavigation)) by getting rid of the buttons and navigating to ViewA in the MainWindow constructor.

Found out that I can navigate successfully if do it when the `Window.Loaded` event fires. Like this...

```
        public MainWindow(IRegionManager regionManager)
        {
            this.regionManager = regionManager;

            // this.Activated += Navigate;       // works
            // this.ContentRendered += Navigate; // works
            // this.Initialized += Navigate;     // no !
            this.Loaded += Navigate;             // works, element laid out, rendered, and ready

            InitializeComponent();
        }

        private void Navigate(object sender, EventArgs e)
        {   
            regionManager.RequestNavigate("ContentRegion", "ViewA", MyNavCallback);
        }
```

Applying this to my own application, it still doesn't work-- probably because I missed something because I am in a dumpster fire of refactoring (my problem).

But before going further down the rabbit hole, is the above is the "easy path" for making the initial Navigation work for MainWindow regions. Is there a better way? 

I realize that Prism does not take an opinionated approach, that's great, but it also means that it's easy to put ~~oneself~~ (myself) in a mess.  By the way, I notice that RequestNavigate does NOT throw exceptions. Instead it sets result=false in NavigationResult and puts the exception, if any, into an Error property. What's the rationale for that? Why would one not want an Exception thrown when the navigation fails? FWIW, in this case Error is null even when there are no Regions yet.

**Comment by brianlagunas** [comment, marked as answer]
Usually, your Views will go in a Module and from within that Module you can then navigate your initial views into the region.

You could also just use view discovery instead of navigation for initial views.

**Reply by angelotrivelli** [reply]
Thanks Brian!

Indeed, if I call `RequestNavigate()` from a Module's `OnInitialized()` it **does** work. 

I got into the problems when I tried to `RequestNavigate()` from within the View or ViewModel of the MainWindow shell. `RegisterViewWithRegion()` does seem to work in that scenario, fwiw.


```
        public MainWindowViewModel(IRegionManager regionManager)
        {
            _regionManager = regionManager;
            Navigate("ViewA");
        }

        private void Navigate(string navigatePath)
        {
            ///                                      this works...
            // if (navigatePath != null)
            //     _regionManager.RegisterViewWithRegion("ContentRegion", typeof(ViewA));


            //                                       this does NOT... 
            //                                       but will work if instead called from Module Initialize() !!
            if (navigatePath != null)
                 _regionManager.RequestNavigate("ContentRegion", "ViewA");
        }

```
